// Copyright 2024 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto2";

package hintless_pir;

import "shell_encryption/rns/serialization.proto";

// For faster allocations of sub-messages.
option cc_enable_arenas = true;

// A LinPIR request sent from the client to the server.
message LinPirRequest {
  // The "b" component of the ciphertext that encrypts the query vector.
  optional rlwe.SerializedRnsPolynomial ct_query_b = 1;

  // The "b" components of the Galois key for homomorphic rotation.
  repeated rlwe.SerializedRnsPolynomial gk_key_bs = 2;
}

// A LinPIR response sent from the server to the client.
message LinPirResponse {
  message EncryptedInnerProduct {
    repeated rlwe.SerializedRnsRlweCiphertext ct_blocks = 1;
  }

  repeated EncryptedInnerProduct ct_inner_products = 1;
}
